### ISSUE TITLE
Fix false positives for styled-components in no-duplicate-selectors

### ISSUE THREAD

**Issue by hudochenkov** [stylelint/stylelint#3863]
> Clearly describe the bug

False positives for `no-duplicate-selectors` in Styled Components. The rule looks into two different components which is not correct. Every styled component should be treated as a separate file if we speak in CSS terms. `Left` and `Right` are two different components, and stylelint can't know if their rules collapse or not.

This is most likely has to do with how `postcss-styled` parses a file.

Related issue #3862.

Test repository: https://github.com/hudochenkov/styled-component-stylelint-bugs/tree/master/false-positive-no-duplicate-selectors

> Which rule, if any, is the bug related to?

`no-duplicate-selectors`

> What CSS is needed to reproduce the bug?

```js
import styled from 'styled-components';
import { ButtonStyled } from '../../button/Button.styled';

export const Left = styled.div`
  a {
    color: red;
  }

  ${ButtonStyled} {
    margin-right: 8px;
  }
`;

export const Right = styled.div`
  a {
    color: blue;
  }

  ${ButtonStyled} {
    margin-left: 8px;
  }
`;
```

> What stylelint configuration is needed to reproduce the bug?

```json
{
    "extends": ["stylelint-config-recommended", "stylelint-config-styled-components"]
}
```

> Which version of stylelint are you using?

`9.9.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint "**/*.styled.js" --syntax styled`

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

Yes, it's related to styled components.

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors did you get)?

The following warnings were flagged:

```shell
Header.styled.js
 15:3  ✖  Unexpected duplicate selector "a", first used at line 5                 no-duplicate-selectors
 19:3  ✖  Unexpected duplicate selector "${ButtonStyled}", first used at line 9   no-duplicate-selectors
```

**Comment by jeddy3** [comment]
> Every styled component should be treated as a separate file if we speak in CSS terms

Interesting. I hadn't really thought of it like that. I recently [answered a SO question](https://stackoverflow.com/questions/53499725/no-duplicate-selectors-error-for-different-selectors-inside-the-same-file) saying that the current behaviour was intended, but now I'm questioning that.

I think it does make sense if postcss-styled parses each component definition as a source.

**Comment by gucong3000** [comment]
I have a temporary solution.
Run `stylelint` as plugin of postcss.
The problem was caused by the fact that CLI of `stylelint` did not use `postcss.process()`.
https://github.com/gucong3000/postcss-syntax/blob/bbc7d9db05f9ffecc36deaebf382fbbcde6ebddb/patch-postcss.js#L11-L19

**Comment by jeddy3** [comment]
> The problem was caused by the fact that CLI of stylelint did not use `postcss.process()`

Can we change stylelint to use `postcss.process` or will there be unintended side effects?

**Comment by jeddy3** [comment]
Fixed in #3875
